docs/lazy-adr/adr-009-state-fraud-proofs.md

```diff
@@ -0,0 +1,233 @@
+
+Overall, State Fraud Proofs enable trust-minimization between full nodes and light clients as long as there is at least one honest full node in the system that will generate state fraud proofs.
+
+![State Fraud Proofs](./figures/state_fraud_proofs.png)
+
+## Alternative Approaches
+
+For light clients to detect invalid blocks and verify it themselves, alternatives include downloading the whole state themselves or using zero-knowledge validity proof systems.
+
+## Detailed Design
+
+### Detecting Fraudulent State Transitions
+
+Rollmint blocks contain a field called `Intermediate State Roots` in block data:
+
+```go
+
+// Data defines rollmint block data.
+type Data struct {
+	Txs                    Txs
+	IntermediateStateRoots IntermediateStateRoots
+	Evidence               EvidenceData
+}
+```
+
+These Intermediate State Roots (ISRs) are initially generated by a Rollmint sequencer during block execution which uses the ABCI interface.
+
+The following ABCI methods are called during block execution:
+`BeginBlock` at the start of a block
+`DeliverTx` for each transaction
+`EndBlock` at the end of a block
+
+After each of the above ABCI method calls, we generate an intermediate state root using a new ABCI method we introduce:
+
+```protobuf
+service ABCIApplication {
+    rpc GetAppHash(RequestGetAppHash) returns (ResponseGetAppHash);
+}
+
+message RequestGetAppHash {}
+
+message ResponseGetAppHash {
+  bytes app_hash = 1;
+}
+```
+
+This `GetAppHash` ABCI method returns an equivalent of `CommitID` hash in the ABCI method `Commit` and thus provides a way to extract ISRs from an app without doing any disk write operations.
+
+Full nodes use these ISRs to detect fraudulent state transitions. A full node must also execute all state transitions (`BeginBlock`, `DeliverTx`, and `EndBlock` calls) and compute its own Intermediate State Roots (ISRs). After each state transition, a full node compares the corresponding ISR with the ISR given by the Sequencer. If it finds a mismatch between its own computed ISR and and one given by the Sequencer, a fraudulent transition is detected and it moves on to generate a Fraud Proof.
+
+### Generating Fraud Proofs
+
+We introduce the following ABCI method to enable Fraud Proof Generation in the Cosmos SDK:
+
+```protobuf
+service ABCIApplication {
+    rpc GenerateFraudProof(RequestGenerateFraudProof) returns (ResponseGenerateFraudProof);
+}
+```
+
+With this new ABCI method, a Rollmint Full Node can send a request to a Cosmos SDK app to generate a Fraud Proof. In this request, it includes a list of all the state transitions from the start of the block upto the fraudulent state transition. The last non-nil state transition in this list corresponds to the fraudulent state transition.
```

Comment:
[non-blocking] I am wondering: would it not also be possible to have the fraud proof logic entirely on the Cosmos SDK side rather than the rollmint side? eg the Cosmos SDK compares the ISRs for incoming blocks. Or would that be even more complicated?

Reply:
I can see 2 issues.
1. Cosmos-SDK is more concerned about transactions and app state than blocks. It's informed about block beginning and ending (`BeginBlock`, `EndBlock`), but have no explicit information about number of transactions in the block, etc. It's the role of rollmint to call `DeliverTx` for each transaction; rollmint drives the communication with cosmos-sdk. Changing this flow would be complicated.
2. The more changes we introduce to cosmos-sdk, the harder it will be to keep a fork up to date.